### ISSUE TITLE
💅 `useValidAnchor` is throwing a false positive for `href="#"`

### ISSUE THREAD

**Issue by Th3S4mur41** [biomejs/biome#10656]
### Environment information

<details>

```bash

CLI:
  Version:                      2.5.0
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_DISTRIBUTION:                npm
  BIOME_LOG_PATH:                    unset
  BIOME_LOG_PREFIX_NAME:             unset
  BIOME_LOG_LEVEL:                   unset
  BIOME_LOG_KIND:                    unset
  BIOME_CONFIG_PATH:                 unset
  BIOME_THREADS:                     unset
  BIOME_WATCHER_KIND:                unset
  BIOME_WATCHER_POLLING_INTERVAL:    unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v24.16.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         npm/11.13.0

Biome Configuration:
  Status:                       Loaded successfully.
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true
  HTML full support enabled:    true

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  true
  GraphQL enabled:              unset
  Recommended:                  unset
  Enabled rules:
    a11y/noAccessKey
    a11y/noAmbiguousAnchorText
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noImportantStyles
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useIndexOf
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noDuplicateArgumentNames
    correctness/noDuplicateAttributes
    correctness/noDuplicateEnumValueNames
    correctness/noDuplicateInputFieldNames
    correctness/noDuplicateVariableNames
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useGraphqlNamedOperations
    correctness/useIsNan
    correctness/useLoneAnonymousOperation
    correctness/useParseIntRadix
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    nursery/useBaseline
    performance/noAccumulatingSpread
    performance/noDynamicNamespaceImportAccess
    security/noBlankTarget
    security/noGlobalEval
    security/noScriptUrl
    style/noDescendingSpecificity
    style/noInferrableTypes
    style/noNonNullAssertion
    style/noParameterAssign
    style/noUnusedTemplateLiteral
    style/noUselessElse
    style/useArrayLiterals
    style/useAsConstAssertion
    style/useConst
    style/useDefaultParameterLast
    style/useDeprecatedReason
    style/useEnumInitializers
    style/useExponentiationOperator
    style/useExportType
    style/useImportType
    style/useLiteralEnumMembers
    style/useNodejsImportProtocol
    style/useNumberNamespace
    style/useShorthandFunctionType
    style/useSingleVarDeclarator
    style/useTemplate
    suspicious/noApproximativeNumericConstant
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noBiomeFirstException
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDeprecatedMediaType
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateEnumValues
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExplicitAny
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noNonNullAssertedOptionalChain
    suspicious/noOctalEscape
    suspicious/noProto
    suspicious/noPrototypeBuiltins
    suspicious/noQuickfixBiome
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noTsIgnore
    suspicious/noUnknownAtRules
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noUselessEscapeInString
    suspicious/noUselessRegexBackrefs
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useBiomeIgnoreFolder
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useIterableCallbackReturn
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0

```

</details>

### Rule name

lint/a11y/useValidAnchor

### Playground link

https://biomejs.dev/playground/?tab=formatter&pane=Diagnostics&code=PABhACAAaAByAGUAZgA9ACIAIwAiACAAYwBsAGEAcwBzAD0AIgB0AG8ALQB0AG8AcAAiAD4AVABvACAAdABoAGUAIAB0AG8AcAA8AC8AYQA%2BAA%3D%3D&language=html

### Expected result

This is a valid anchor to scroll to the top of the page and should not throw an error

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
Not a bug. The JSX rule has the same behaviour, and only `#` **isn't** a valid URL. To have users navigating at the top of the page, you have to use a hidden ID, which has a the top of the URL. Check how starlight does it

**Comment by Th3S4mur41** [comment]
@ematipico  
Actually, according to the official HTML Living Standard, using `href="#"` to scroll to the top of the page is fully valid, explicitly specified behavior. It is not a bug, nor is a hidden ID required.

Here is exactly how the specification defines it:

1. **The URL is Valid:** According to the URL standard, a `#` on its own represents a URL with an **empty fragment** identifier.

2. **The Processing Spec:** In the HTML Living Standard (Section: [7.4.6.4 Scrolling to a fragment](https://html.spec.whatwg.org/multipage/browsing-the-web.html#scrolling-to-a-fragment)), the browser is given an explicit algorithm to determine the target element. The spec states:
*"If fragment is the empty string, then return the special value top of the document."*

3. **The Scrolling Spec:** The standard then dictates that if the designated target is the top of the document, the browser must:  
*"Scroll to the beginning of the document."*

While using a hidden ID (like `<div id="top">`) or looking at how frameworks like Starlight handle it are perfectly fine architectural choices for explicit routing, they are technically redundant for this specific utility. The browser is natively mandated by the web specs to scroll to the top when encountering an empty fragment (`#`).

**Comment by ematipico** [comment]
Is there a link to a resource that  says "# is accessible"? The WHATWG spec isn't the accessibility spec. We're not debating if `#` is valid, we're debating if it's accessible

**Comment by Th3S4mur41** [comment]
> Is there a link to a resource that says "# is accessible"? The WHATWG spec isn't the accessibility spec. We're not debating if `#` is valid, we're debating if it's accessible

Aah, got it... In that case you are absolutely correct.
It does indeed create a disconnect between scrolling and focus on some browser/screen reader combination (I didn't know about).
I found some discussions about this and have just verified it as well with Edge and Narrator/NVDA in a quick test.

So this should indeed throw an error, but I think the explanation needs some improvement to point to the actual accessibility problem with this.

So far, I couldn't find a proper documentation for this though
[WebAIM](https://webaim.org/techniques/hypertext/hypertext_links#same_page) doesn't mention it and only references its use in combination with javascript click handler.

Same with [ESLint a11Y plugin](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-is-valid.md#case-i-want-navigable-links), it mentions `#` is not a valid `href` value, which is not correct, but doesn't mention anywhere that using it is actually creating an accessibility issue.

So while it could be inferred that the problem is with accessibility, because of the origin of the rule, the documentation itself is confusing at best.

So maybe we could add something like this explicitly to the docs?

While href="#" is structurally valid in the HTML parser specification for shifting the visual viewport, it presents an active Focus Management failure for keyboard and assistive technologies.

1. **Focus Desynchronization:** An empty hash (#) does not reference a specific element ID node. When activated, the browser scrolls the page visually, but might (some user-agent) leave the keyboard focus ring pinned down to the trigger element in the footer. The next Tab keystroke pulls the user right back down to the bottom.

2. **Spec Compliance:** This directly breaks WCAG 2.4.3 (Focus Order) because the physical reading sequence and focus structure fall completely out of sync.

**Comment by ematipico** [comment]
The docs of the lint rule already talk about it, and the diagnostics already provide a link to the explanation. I'm not sure what else you need. If you think there's else is needed, feel free to send a PR, however I believe the docs and the diagnostics are already enough.
